crates/wysiwyg/src/composer_model/menu_action.rs

```diff
@@ -0,0 +1,91 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    dom::{
+        unicode_string::{UnicodeStr, UnicodeStringExt},
+        Range,
+    },
+    ComposerModel, MenuAction, PatternKey, SuggestionPattern, TrailingStrategy,
+    UnicodeString,
+};
+
+impl<S> ComposerModel<S>
+where
+    S: UnicodeString,
+{
+    /// Compute the menu action for current composer model state.
+    pub(crate) fn compute_menu_action(&self) -> MenuAction {
+        let (s, e) = self.safe_selection();
+        let range = self.state.dom.find_range(s, e);
+        if range.locations.iter().any(|l| l.kind.is_code_kind()) {
+            return MenuAction::None;
+        }
+        let (raw_text, start, end) = self.extended_text(range);
+        if let Some((key, text)) = Self::pattern_for_text(raw_text) {
+            MenuAction::Suggestion(SuggestionPattern {
+                key,
+                text,
+                start,
+                end,
+                trailing_strategy: TrailingStrategy::default(),
+            })
+        } else {
+            MenuAction::None
+        }
+    }
+
+    /// Compute extended text from a range. Text is extended up
+    /// to the leading/trailing of the text nodes, or up to the
+    /// first whitespace found.
+    /// Returns the extended text, and its start/end locations.
+    fn extended_text(&self, range: Range) -> (S, usize, usize) {
+        range
+            .leaves()
+            .filter_map(|loc| {
+                self.state
+                    .dom
+                    .lookup_node(&loc.node_handle)
+                    .as_text()
+                    .map(|t| (t, loc.start_offset..loc.end_offset))
+            })
+            .fold(
+                (S::default(), range.start(), range.end()),
+                |(mut text, s, e), (t, range)| {
+                    let (node_text, start_offset, end_offset) =
+                        t.extended_text_for_range(range);
+                    text.push(node_text);
+                    (text, s - start_offset, e + end_offset)
+                },
+            )
+    }
+
+    /// Compute at/hash/slash pattern for a given text.
+    /// Return pattern key and associated text, if it exists.
+    fn pattern_for_text(mut text: S) -> Option<(PatternKey, String)> {
+        let Some(first_char) = text.pop_first() else {
+            return None;
+        };
+        let Some(key) = PatternKey::from_char(first_char) else {
+            return None;
+        };
+
+        // Exclude if there is inner whitespaces.
+        if text.chars().any(|c| matches!(&c, ' ' | '\x09'..='\x0d')) {
```

Comment:
Could it be safer to use a built-in Rust utility for detecting whitespace? I saw [`is_whitespace`](https://doc.rust-lang.org/std/primitive.char.html#method.is_whitespace) and the source [list of characters](https://www.unicode.org/Public/UCD/latest/ucd/PropList.txt) that mention a few more characters than the range of `'\x09'..='\x0d'`.

If not, perhaps we can extract this to a named utility that explains why we need to do it in a custom way?

Reply:
Yeah actually I thought about the same and did not commit the change. Right now we are using this matches macro for both this feature and the backspace/delete word feature, and in both cases I assume it is fair to stop at any character that has the unicode `whitespace` property.

Reply:
I replaced usages of the macro here: https://github.com/matrix-org/matrix-rich-text-editor/pull/591/commits/248077dffa77f09d6ccd176c39c71b16c6c64220

Looking at Alun's helper, I wondered if we cared about Punctuation signs, but at the moment it would break things to try to care about them, because `@` is considered one. And I guess it's fair that a display name could contain some punctuation signs for some reason.